File: Benchmarks_FSharp/Core/Systems/InputSystem.fs
namespace Benchmarks.Core.Systems

open System.Collections.Generic
open Benchmarks.Core.Components
open Benchmarks.Core.Entity
open Benchmarks.Core.Scene
open Benchmarks.Core.Systems.Contract
open Benchmarks.Core.Types

type InputSystem() =
    inherit ISystem()

    let mutable eventManager: EventManager = Unchecked.defaultof<EventManager>
    let _pressedKeys = HashSet<int>([ 0; 1; 2; 3 ])
    let _mousePosition = Vector2(500f, 400f)

    override this.InitializeSystem(entities: EntityManager) =
        eventManager <- EventManager.Instance
        this.EntityManager <- entities

    override this.UpdateSystem(gameTime: GameTime) =
        if isNull this.EntityManager then ()
        else
        let inputComponents = this.GetInputComponentsAsSpan()
        let entities = this.EntityManager.YotsubaEntities.AsSpan()
        for i = 0 to entities.Length - 1 do
            let entity = &entities.[i]
            if entity.HasComponent(YTBComponent.Input) then
                let inputComponent = &inputComponents.[entity.Id]
                InputSystem.ProcessKeyboard(_pressedKeys, entity.Id, &inputComponent, gameTime)
                InputSystem.ProcessMouse(_mousePosition, entity.Id, &inputComponent, gameTime)

    static member private ProcessKeyboard(pressedKeys: HashSet<int>, entityId: int, inputComponent: byref<InputComponent>, gameTime: GameTime) =
        if not (inputComponent.HasInput(InputInUse.HasKeyboard)) then ()
        elif isNull inputComponent.KeyBoard then ()
        else
            for mapping in inputComponent.KeyBoard do
                if pressedKeys.Contains(mapping.Value) then
                    mapping.Key |> ignore

    static member private ProcessMouse(mousePosition: Vector2, entityId: int, inputComponent: byref<InputComponent>, gameTime: GameTime) =
        if not (inputComponent.HasInput(InputInUse.HasMouse)) then ()
        elif isNull inputComponent.Mouse then ()
        else
            for _mapping in inputComponent.Mouse do
                mousePosition |> ignore

    override _.SharedEntityForEachUpdate(_, _) = ()
    override _.SharedEntityInitialize(_) = ()
    override _.Dispose() = ()

File: Benchmarks/Core/Systems/InputSystem.cs
using Benchmarks.Core.Components;
using Benchmarks.Core.Entity;
using Benchmarks.Core.Scene;
using Benchmarks.Core.Systems.Contract;
using Benchmarks.Core.Types;

namespace Benchmarks.Core.Systems
{
    /// <summary>
    /// Sistema encargado de manejar la entrada del usuario.
    /// Simula el procesamiento de input con dictionary lookups.
    /// </summary>
    public class InputSystem : ISystem
    {
        private EventManager EventManager { get; set; }

        // Simulated input state
        private readonly HashSet<int> _pressedKeys = new() { 0, 1, 2, 3 };
        private readonly Vector2 _mousePosition = new(500, 400);

        public override void InitializeSystem(EntityManager @entities)
        {
            EventManager = EventManager.Instance;
            EntityManager = @entities;
        }

        public override void UpdateSystem(GameTime gameTime)
        {
            if (EntityManager == null) return;

            Span<InputComponent> inputComponents = GetInputComponentsAsSpan();

            foreach (ref Yotsuba entity in EntityManager.YotsubaEntities.AsSpan())
            {
                if (entity.HasComponent(YTBComponent.Input))
                {
                    ref InputComponent inputComponent = ref inputComponents[entity.Id];
                    ProcessKeyboard(entity.Id, ref inputComponent, gameTime);
                    ProcessMouse(entity.Id, ref inputComponent, gameTime);
                }
            }
        }

        private void ProcessKeyboard(int entityId, ref InputComponent inputComponent, GameTime gameTime)
        {
            if (!inputComponent.HasInput(InputInUse.HasKeyboard)) return;
            if (inputComponent.KeyBoard == null) return;

            foreach (var mapping in inputComponent.KeyBoard)
            {
                if (_pressedKeys.Contains(mapping.Value))
                {
                    // Simulate key processing (equivalent to publishing events)
                    _ = mapping.Key;
                }
            }
        }

        private void ProcessMouse(int entityId, ref InputComponent inputComponent, GameTime gameTime)
        {
            if (!inputComponent.HasInput(InputInUse.HasMouse)) return;
            if (inputComponent.Mouse == null) return;

            foreach (var mapping in inputComponent.Mouse)
            {
                // Simulate mouse processing
                _ = _mousePosition;
            }
        }

        public override void SharedEntityForEachUpdate(ref Yotsuba Entidad, GameTime time) { }
        public override void SharedEntityInitialize(ref Yotsuba Entidad) { }
        public override void Dispose() { }
    }
}
